compiler/rustc_mir_transform/src/const_prop_lint.rs

```diff
@@ -485,18 +485,30 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     .and_then(|op| self.ecx.read_immediate(&op).ok())
                     .map_or(DbgVal::Underscore, |op| DbgVal::Val(op.to_const_int()))
             };
-            let msg = match msg {
-                AssertKind::DivisionByZero(op) => AssertKind::DivisionByZero(eval_to_int(op)),
-                AssertKind::RemainderByZero(op) => AssertKind::RemainderByZero(eval_to_int(op)),
+            let (msg, emit_overflow_lint) = match msg {
+                AssertKind::DivisionByZero(op) => {
+                    (AssertKind::DivisionByZero(eval_to_int(op)), false)
+                }
+                AssertKind::RemainderByZero(op) => {
+                    (AssertKind::RemainderByZero(eval_to_int(op)), false)
+                }
                 AssertKind::Overflow(bin_op @ (BinOp::Div | BinOp::Rem), op1, op2) => {
                     // Division overflow is *UB* in the MIR, and different than the
                     // other overflow checks.
-                    AssertKind::Overflow(*bin_op, eval_to_int(op1), eval_to_int(op2))
+                    (AssertKind::Overflow(*bin_op, eval_to_int(op1), eval_to_int(op2)), false)
+                }
+                AssertKind::Overflow(bin_op @ (BinOp::Shl | BinOp::Shr), op1, op2) => {
+                    // A hack that fixes #117949
+                    // Ideally check_binary_op() should check these shift ops,
+                    // but it can't because they are getting removed from the MIR
+                    // during const promotion. So we check the associated asserts
+                    // here instead as they are not removed by promotion.
+                    (AssertKind::Overflow(*bin_op, eval_to_int(op1), eval_to_int(op2)), true)
```

Comment:
If you emit the lint here then you should be able to remove the shift handling from check_binary_op. That should also fix the duplicate lints.

Reply:
I tried that, but it eliminates only 6 of the 59 new duplicates added by this change.

Reply:
What exactly did you try? I should look something like this:
```diff
diff --git a/compiler/rustc_mir_transform/src/const_prop_lint.rs b/compiler/rustc_mir_transform/src/const_prop_lint.rs
index 99eecb567f2..d36798518e2 100644
--- a/compiler/rustc_mir_transform/src/const_prop_lint.rs
+++ b/compiler/rustc_mir_transform/src/const_prop_lint.rs
@@ -317,37 +317,9 @@ fn check_binary_op(
         });
         let l = self
             .use_ecx(location, |this| this.ecx.read_immediate(&this.ecx.eval_operand(left, None)?));
-        // Check for exceeding shifts *even if* we cannot evaluate the LHS.
+        // Shifts are linted via the `Assert` terminator, not this function.
         if matches!(op, BinOp::Shr | BinOp::Shl) {
-            let r = r.clone()?;
-            // We need the type of the LHS. We cannot use `place_layout` as that is the type
-            // of the result, which for checked binops is not the same!
-            let left_ty = left.ty(self.local_decls(), self.tcx);
-            let left_size = self.ecx.layout_of(left_ty).ok()?.size;
-            let right_size = r.layout.size;
-            let r_bits = r.to_scalar().to_bits(right_size).ok();
-            if r_bits.is_some_and(|b| b >= left_size.bits() as u128) {
-                debug!("check_binary_op: reporting assert for {:?}", location);
-                let source_info = self.body().source_info(location);
-                let panic = AssertKind::Overflow(
-                    op,
-                    match l {
-                        Some(l) => l.to_const_int(),
-                        // Invent a dummy value, the diagnostic ignores it anyway
-                        None => ConstInt::new(
-                            ScalarInt::try_from_uint(1_u8, left_size).unwrap(),
-                            left_ty.is_signed(),
-                            left_ty.is_ptr_sized_integral(),
-                        ),
-                    },
-                    r.to_const_int(),
-                );
-                self.report_assert_as_lint(
-                    source_info,
-                    AssertLint::ArithmeticOverflow(source_info.span, panic),
-                );
-                return None;
-            }
+            return None;
         }
 
         if let (Some(l), Some(r)) = (l, r) {
```

Reply:
Oh. I just got rid of the entire `if` block. Didn't realize I must return `None` to disable shift handling.

I have done that now in the latest commit I pushed. It has eliminated all duplicates. However, we have the opposite problem now. Many of the shift lints in release builds have been silenced possibly because we don't emit asserts in those builds.